Bug fix for source interval method

ckarwin opened the pull request:
This PR fixes issue #299. When I added the livetime to the spacecraft file, I didn't update the source interval method, which essentially makes a sub-selection of the ori file. That has now been updated. I tested it by running the following: 

```
from cosipy.spacecraftfile import SpacecraftFile
from astropy.time import Time

ori_file = "DC3_final_530km_3_month_with_slew_15sbins_GalacticEarth_SAA.ori"
ori = SpacecraftFile.parse_from_file(ori_file)
new_ori = ori.source_interval(Time(1842123855.0221176, format = 'unix'), Time(1842123887.7090132, format = 'unix'))

print("original ori start:")
print(ori._time[0].value)
print(ori._time[1].value)
print(ori.livetime[0])

print("new ori start")
print(new_ori._time[0].value)
print(new_ori._time[1].value)
print(new_ori.livetime[0])

print("original ori end:")
print(ori._time[-1].value)
print(ori._time[-2].value)
print(ori.livetime[-1])

print("new ori end")
print(new_ori._time[-1].value)
print(new_ori._time[-2].value)
print(new_ori.livetime[-1])
```

Which gives this:

```
original ori start:
1835487300.0
1835487315.0
15.0


new ori start
1842123855.0221176
1842123870.0
14.977882385253906


original ori end:
1843467255.0
1843467240.0
0.0


new ori end
1842123887.7090132
1842123885.0
2.7090132236480713
```

So it seems to work as expected, but please test for yourself.

I also made a small change to the UnBinned class. For the function that parses tra files, I added a check to the photon event data, ensuring that all the needed info is present in the tra file. Currently, it only checks that each photon event has at least two hits in the detector. If not, the event is not included in the photon list. This is an issue that came up when parsing the primary proton backgrounds for DC3. It only occured for 4 out of millions of events, and so it's very rare.  

@eneights This addressed the issue that you found a couple of weeks ago. If you can test this PR it would be much appreciated!

ckarwin commented:
Note: Once approved, this should also be merged into branch v3.0.x (and eventually main).

eneights commented:
I tested out the source interval selection, and it's working for me!

ckarwin commented:
@eneights Thanks for checking! Are you ok with approving/merging the PR?

eneights commented:
I tried to generate a scatt map using the updated source interval selection, and I'm getting this error:
```
File ~/work/COSI/data-challenges/data_challenge_3/cosipy_fork/cosipy/spacecraftfile/SpacecraftFile.py:611, in SpacecraftFile.get_scatt_map(self, target_coord, nside, scheme, coordsys, r_earth, earth_occ)
    608 weight = self.livetime*u.s
    610 if earth_occ == True:
--> 611     weight[earth_occ_index[:-1]] = 0        
    613 # Fill histogram:
    614 h_ori.fill(x, y, weight = weight)

File ~/software/miniforge3/envs/cosipy-fork/lib/python3.10/site-packages/astropy/units/quantity.py:1331, in Quantity.__setitem__(self, i, value)
   1329 if not self.isscalar and "info" in self.__dict__:
   1330     self.info.adjust_indices(i, value, len(self))
-> 1331 self.view(np.ndarray).__setitem__(i, self._to_own_unit(value))

IndexError: boolean index did not match indexed array along dimension 0; dimension is 2 but corresponding boolean dimension is 1
```

This is the code I ran to get this:
```
sc_orientation = SpacecraftFile.parse_from_file('DC3_final_530km_3_month_with_slew_1sbins_GalacticEarth_SAA.ori')
ori = sc_orientation.source_interval(Time(1835493492.2, format = 'unix'), Time(1835493492.8, format = 'unix'))
source_direction = SkyCoord(l=23.53, b=-53.44, frame='galactic', unit=u.deg)
scatt_map = ori.get_scatt_map(source_direction, nside=response.nside*2, coordsys='galactic')
```

eneights commented:
The energy selection is working for me, so I think once the scatt map issue is fixed, it's ready to be merged. 

I notice a check is failing; is that okay or does that also need to be fixed before merging?

ckarwin commented:
Thank, @eneights. I fixed the scatt_map method. There was a bug in how I was defining the end points for the live time sub-selection in the source interval method. I also added unit tests for the energy selection method. Can you please try again?

ckarwin commented:
I tested the scatt_map fix by adding the following to the code I posted above:
```
source_direction = SkyCoord(l=23.53, b=-53.44, frame='galactic', unit=u.deg)
scatt_map = new_ori.get_scatt_map(source_direction, nside=8, coordsys='galactic')
```
It worked ok for me.

eneights reviewed (approved):
The scatt map works for me now, and everything else looks good!
